src/index.tsx

```diff
@@ -1,71 +1,24 @@
-import React, {
-  ComponentType,
-  ComponentPropsWithoutRef,
-  PropsWithChildren,
-  ReactElement,
-} from 'react';
-
-type ProviderComponent<T> = ComponentType<T>;
-type ProviderComponentProps<T> = ComponentPropsWithoutRef<ProviderComponent<T>>;
-
-type ProviderWithProps<T = unknown> = [
-  ComponentType<T>,
-  ComponentPropsWithoutRef<ComponentType<T>>,
-];
-
-function curryProviders(
-  provider: ProviderComponent<unknown> | ProviderWithProps,
-): ProviderComponent<unknown> {
-  if (!Array.isArray(provider)) {
-    return provider;
-  }
-
-  const [Component, props] = provider;
-
-  return function CurriedComponent({
-    children,
-  }: PropsWithChildren<unknown>): ReactElement {
-    return <Component {...props}>{children}</Component>;
-  };
-}
-
-function nestProviders(
-  PreviousProviders: ProviderComponent<unknown>,
-  CurrentProvider: ProviderComponent<unknown>,
-): ProviderComponent<unknown> {
-  return function NestedProviders({
-    children,
-  }: PropsWithChildren<unknown>): JSX.Element {
-    return (
-      <PreviousProviders>
-        <CurrentProvider>{children}</CurrentProvider>
-      </PreviousProviders>
-    );
-  };
-}
+import React, { PropsWithChildren, ReactElement } from 'react';
+import { nestProviders } from './aggregators/nest-providers';
+import { unWrapTupleProvider } from './aggregators/unwrap-tuple-providers';
+import { ProviderComponent, TupleProviderWithProps } from './types/provider';
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-type Providers<T = any> = Array<ProviderComponent<T> | ProviderWithProps<T>>;
-
-type FlatProvidersProps = PropsWithChildren<{
-  // eslint-disable-next-line @typescript-eslint/member-delimiter-style
-  providers: Providers;
-}>;
+export type Providers<T = any> = Array<
+  ProviderComponent<T> | TupleProviderWithProps<T>
+>;
 
 export default function FlatProviders({
```

Comment:
One more thing here: Is it okay if we do not use default exports?
They make it harder for CJS Environments to import from, and I think it's always better to have so called named exports.

The difference in using would be like this:

```
// import with default export
import FlatProviders from 'react-flat-providers'

// import with named export
import { FlatProviders } from 'react-flat-providers'
```

This would also prevent a warning that is currently thrown on linting as we mix named and default exports here.

Reply:
Yeah, that's a good idea, I'll do it in the next commit.